fastcrypto/src/groups/bls12381.rs

```diff
@@ -135,6 +138,43 @@ impl GroupElement for G1Element {
     }
 }
 
+impl Pair for G1Element {
+    type Other = G2Element;
+    type Output = GTElement;
+
+    fn pair(&self, other: &Self::Other) -> Self::Output {
+        let mut self_affine = blst_p1_affine::default();
+        let mut other_affine = blst_p2_affine::default();
+        unsafe {
+            blst_p1_to_affine(&mut self_affine, &self.0);
+            blst_p2_to_affine(&mut other_affine, &other.0);
+        }
+        // The API requires DST although we don't use hash-to-curve here.
+        let unused_dst = [0u8; 3];
+        let mut pairing_blst = Pairing::new(false, &unused_dst);
+        pairing_blst.raw_aggregate(&other_affine, &self_affine);
+        Self::Output::from(pairing_blst.as_fp12())
+    }
```

Comment:
If you are going to compute the pairing with only one set of operands involved, the following is just as efficient and unencumbered by signing APIs:
```
        blst_p1_to_affine(&mut self_a, &self);
        blst_p2_to_affine(&mut other_a, &other);

        blst_miller_loop(&mut gt_point, &other_a, &self_a);
        blst_final_exp(&mut gt_point, &gt_point);
```

Reply:
nice, thanks for the suggestion